DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/register_map_and_streaming_interfaces/src/register_map_and_streaming_interfaces.cpp

```diff
@@ -0,0 +1,209 @@
+ValueT SomethingComplicated(ValueT val) { return (ValueT)(val * (val + 1)); }
+
+/////////////////////////////////////////
+
+struct FunctorStreamingControlIP {
+  // Use the 'conduit' annotation on a kernel argument to specify it to be
+  // a streaming kernel argument.
+  conduit ValueT *input;
+  conduit ValueT *output;
+  // Without the annotations, kernel arguments will be inferred to be streaming
+  // kernel arguments if the kernel control interface is streaming, and
+  // vise-versa.
+  size_t n;
+  FunctorStreamingControlIP(ValueT *in_, ValueT *out_, size_t N_)
+      : input(in_), output(out_), n(N_) {}
+  // Use the 'streaming_interface' annotation on a kernel to specify it to be
+  // a kernel with streaming kernel control signals.
+  streaming_interface void operator()() const {
+    for (int i = 0; i < n; i++) {
+      output[i] = SomethingComplicated(input[i]);
+    }
+  }
+};
+
+struct FunctorRegisterMapControlIP {
+  // Use the 'register_map' annotation on a kernel argument to specify it to be
+  // a register map kernel argument.
+  register_map ValueT *input;
+  // Without the annotations, kernel arguments will be inferred to be register
+  // map kernel arguments if the kernel control interface is register map, and
+  // vise-versa.
+  ValueT *output;
+  // A kernel with register map control can also independently have streaming
+  // kernel arguments, when annotated by 'conduit'.
+  conduit size_t n;
+  FunctorRegisterMapControlIP(ValueT *in_, ValueT *out_, size_t N_)
+      : input(in_), output(out_), n(N_) {}
+  register_map_interface void operator()() const {
+    for (int i = 0; i < n; i++) {
+      output[i] = SomethingComplicated(input[i]);
+    }
+  }
+};
+
+void TestLambdaRegisterMapControlKernel(sycl::queue &q, ValueT *in, ValueT *out,
+                                        size_t count) {
+  // In the Lambda programming model, all kernel arguments will have the same
+  // interface as the kernel control interface.
+  q.single_task<LambdaRegisterMapControlIP>([=] register_map_interface {
+     for (int i = 0; i < count; i++) {
+       out[i] = SomethingComplicated(in[i]);
+     }
+   }).wait();
+
+  std::cout << "\t Done" << std::endl;
+}
+
+void TestLambdaStreamingControlKernel(sycl::queue &q, ValueT *in, ValueT *out,
+                                      size_t count) {
+  // In the Lambda programming model, all kernel arguments will have the same
+  // interface as the kernel control interface.
+  q.single_task<LambdaStreamingControlIP>([=] streaming_interface {
+     for (int i = 0; i < count; i++) {
+       out[i] = SomethingComplicated(in[i]);
+     }
+   }).wait();
+
+  std::cout << "\t Done" << std::endl;
+}
+
+template <typename KernelType>
+void TestFunctorKernel(sycl::queue &q, ValueT *in, ValueT *out, size_t count) {
+  q.single_task(KernelType{in, out, count}).wait();
+
+  std::cout << "\t Done" << std::endl;
+}
+
+int main(int argc, char *argv[]) {
+#if defined(FPGA_EMULATOR)
```

Comment:
I recommend defaulting to the emulator instead:

```
#if FPGA_SIMULATOR
    std::cout << "using FPGA Simulator." << std::endl;
    sycl::ext::intel::fpga_simulator_selector device_selector;
#elif FPGA_HARDWARE
    std::cout << "using FPGA Hardware." << std::endl;
    sycl::ext::intel::fpga_selector device_selector;
#else  // #if FPGA_EMULATOR
    std::cout << "using FPGA Emulator." << std::endl;
    sycl::ext::intel::fpga_emulator_selector device_selector;
#endif
```

This means that a user could compile successfully with 

```
icpx -fsycl register_map_and_streaming_interfaces.cpp
```

Reply:
i agree with that.

Reply:
Yeah this is why we'll have to update the selectors